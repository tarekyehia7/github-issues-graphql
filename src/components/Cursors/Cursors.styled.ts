import styled from 'styled-components';

export const CursorLink = styled.button`
    color: #0969da;
    min-width: 32px;
    padding: 5px 10px;
    font-style: normal;
    line-height: 20px;
    text-align: center;
    white-space: nowrap;
    vertical-align: middle;
    cursor: pointer;
    -webkit-user-select: none;
    user-select: none;
    border: 1px solid transparent;
    border-radius: 6px;
    transition: border-color .2s cubic-bezier(0.3, 0, 0.5, 1);
    background-color: transparent;
    &:disabled {
        color: #8c959f;
        cursor: default;
        border-color: transparent;
        &: hover {
            border-color: transparent;
        }
    }
    &:hover {
        text-decoration: none;
        border-color: #d0d7de;
        transition-duration: .1s;
    }
`;

export const LinksContainer = styled.div`
    display: flex;
    padding: 1rem;
    justify-content: space-evenly;
    align-items: center;
`;