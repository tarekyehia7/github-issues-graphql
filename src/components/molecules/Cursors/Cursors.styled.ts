import styled from 'styled-components';

export const LinksContainer = styled.div`
	display: flex;
	padding: ${({ theme }) => theme.margins.margin4};
	justify-content: space-evenly;
	align-items: center;
`;
