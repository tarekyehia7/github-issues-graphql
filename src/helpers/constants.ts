export const constants = {
    repository: 'facebook',
    projectName: 'react',
	searchTerm: 'title,body',
    issuesPerPage: 10,
    githubUrl: 'https://github.com'
};