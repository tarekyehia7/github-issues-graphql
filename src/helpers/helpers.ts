import formatDistanceToNowStrict from 'date-fns/formatDistanceToNowStrict';

export const getPagesNumber = (totalIssuesCount: number, issuesPerPage: number): number => {
	const totalPages = Math.ceil(totalIssuesCount / issuesPerPage);
	return !totalPages ? 1 : totalPages;
};

export const formatDate = (date: string) => {
	try {
		const myDate = new Date(date);
		const myFormattedDate = formatDistanceToNowStrict(myDate, { addSuffix: true });
		return myFormattedDate;
	} catch (e) {
		console.error(e);
		return '';
	}
};
