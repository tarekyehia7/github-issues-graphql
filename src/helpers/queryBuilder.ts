import { StatusEnum } from "../pages/Issues/Issues";

interface Query {
    repo: string;
    input: string;
    is: string;
    in: string;
    status: StatusEnum;
}

export const buildQuery = (query: Query) => {
    return Object.entries(query)
        .reduce((str, [p, val]) => `${str}${p}:${val} `, '')
        .replace('status:', '')
        .replace('input:', '')
        .trim();
};