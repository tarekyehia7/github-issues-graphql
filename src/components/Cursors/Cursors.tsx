import React from 'react';

import { CursorLink, LinksContainer } from './Cursors.styled';

type CursorsProps = {
    previousButtonDisabled: boolean;
    nextButtonDisabled: boolean;
    pageNumber: number;
    totalPages: number;
    loadNextData: () => void;
    loadPreviousData: () => void;
};

const Cursors = ({ loadNextData, previousButtonDisabled, nextButtonDisabled, pageNumber, totalPages, loadPreviousData }: CursorsProps) => {
    
    
    return (
        <LinksContainer>
            <CursorLink
                onClick={loadPreviousData}
                disabled={previousButtonDisabled}
            >
                {`< Previous`}
            </CursorLink>
            <div>
                Page {pageNumber} of {totalPages}
            </div>
            <CursorLink
                onClick={loadNextData}
                disabled={nextButtonDisabled}
            >
                {`Next >`}
            </CursorLink>
        </LinksContainer>
    );
};

export default Cursors