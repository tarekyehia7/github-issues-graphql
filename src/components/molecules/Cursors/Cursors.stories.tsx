import React, { useEffect, useState } from 'react';
import { ComponentStory, ComponentMeta } from '@storybook/react';

import Cursors, { CursorsProps } from './Cursors';

export default {
	title: 'molecules/Cursors',
	component: Cursors,
	argTypes: {
		loadNextData: { action: 'clicked' },
		loadPreviousData: { action: 'clicked' },
	},
} as ComponentMeta<typeof Cursors>;

export const Primary: ComponentStory<typeof Cursors> = (args: CursorsProps) => {
	const [pageNumber, setPageNumber] = useState(args.pageNumber);

	useEffect(() => {
		setPageNumber(args.pageNumber);
	}, [args.pageNumber]);

	const loadNextData = () => {
		if (pageNumber <= args.totalPages) {
			setPageNumber(pageNumber + 1);
		}
	};

	const loadPreviousData = () => {
		if (pageNumber > 1) {
			setPageNumber(pageNumber - 1);
		}
	};

	return (
		<Cursors
			pageNumber={pageNumber}
			loadNextData={loadNextData}
			loadPreviousData={loadPreviousData}
			totalPages={args.totalPages}
			previousButtonDisabled={args.previousButtonDisabled}
			nextButtonDisabled={args.nextButtonDisabled}
		/>
	);
};

Primary.args = {
	pageNumber: 1,
	totalPages: 100,
	previousButtonDisabled: false,
	nextButtonDisabled: false,
};
