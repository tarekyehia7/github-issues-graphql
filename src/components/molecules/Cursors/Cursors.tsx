import React from 'react';

import { Button, ButtonType } from '../../atoms/button/Button';
import { LinksContainer } from './Cursors.styled';

type CursorsProps = {
	previousButtonDisabled: boolean;
	nextButtonDisabled: boolean;
	pageNumber: number;
	totalPages: number;
	loadNextData: () => void;
	loadPreviousData: () => void;
};

const Cursors = ({
	loadNextData,
	previousButtonDisabled,
	nextButtonDisabled,
	pageNumber,
	totalPages,
	loadPreviousData,
}: CursorsProps) => {
	return (
		<LinksContainer>
			<Button
				value={`< Previous`}
				onClick={loadPreviousData}
				disabled={previousButtonDisabled}
				type={ButtonType.Cursor}
			/>
			<div>
				Page {pageNumber} of {totalPages}
			</div>
			<Button
				value={`Next >`}
				onClick={loadNextData}
				disabled={nextButtonDisabled}
				type={ButtonType.Cursor}
			/>
		</LinksContainer>
	);
};

export default Cursors;
