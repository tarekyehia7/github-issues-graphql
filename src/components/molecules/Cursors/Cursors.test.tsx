import React from 'react';
import { fireEvent, render } from '@testing-library/react';
import 'jest-styled-components';

import { Cursors, CursorsProps } from './Cursors';
import { PageWithTheme } from '../../../helpers/testing/helpers';

const mockNextClick = jest.fn();
const mockPreviousClick = jest.fn();

const props: CursorsProps = {
	pageNumber: 1,
	totalPages: 100,
	previousButtonDisabled: false,
	nextButtonDisabled: false,
	loadNextData: mockNextClick,
	loadPreviousData: mockPreviousClick,
};

const renderPage = () => {
	const { container, getByText } = render(
		<PageWithTheme>
			<Cursors {...props} />
		</PageWithTheme>,
	);

	return {
		container,
		getByText,
	};
};

describe('<Cursors />', () => {
	it('Should matches snapshot', async () => {
		const { container } = renderPage();
		expect(container).toMatchSnapshot();
	});

	it('Should click previous button', async () => {
		const { getByText } = renderPage();

		fireEvent.click(getByText('< Previous'));
		expect(mockPreviousClick.mock.calls.length).toBe(1);
	});

	it('Should click next button', async () => {
		const { getByText } = renderPage();

		fireEvent.click(getByText('Next >'));
		expect(mockNextClick.mock.calls.length).toBe(1);
	});
});
