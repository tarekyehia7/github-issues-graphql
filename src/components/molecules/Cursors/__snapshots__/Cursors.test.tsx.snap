// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Cursors /> Should matches snapshot 1`] = `
.c1 {
  color: #0969da;
  min-width: 2rem;
  padding: 5px 10px;
  font-style: normal;
  line-height: 20px;
  text-align: center;
  white-space: nowrap;
  vertical-align: middle;
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  border: 1px solid transparent;
  border-radius: 6px;
  -webkit-transition: border-color 0.2s cubic-bezier(0.3,0,0.5,1);
  transition: border-color 0.2s cubic-bezier(0.3,0,0.5,1);
  background-color: transparent;
  color: #0969da;
  min-width: 2rem;
  padding: 5px 10px;
  font-style: normal;
  line-height: 20px;
  text-align: center;
  white-space: nowrap;
  vertical-align: middle;
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  border: 1px solid transparent;
  border-radius: 6px;
  -webkit-transition: border-color 0.2s cubic-bezier(0.3,0,0.5,1);
  transition: border-color 0.2s cubic-bezier(0.3,0,0.5,1);
  background-color: transparent;
}

.c1:disabled {
  color: #8c959f;
  cursor: default;
  border-color: transparent;
}

.c1:disabled:hover {
  border-color: transparent;
}

.c1:hover {
  -webkit-text-decoration: none;
  text-decoration: none;
  border-color: #d0d7de;
  -webkit-transition-duration: 0.1s;
  transition-duration: 0.1s;
}

.c1:disabled {
  color: #8c959f;
  cursor: default;
  border-color: transparent;
}

.c1:disabled:hover {
  border-color: transparent;
}

.c1:hover {
  -webkit-text-decoration: none;
  text-decoration: none;
  border-color: #d0d7de;
  -webkit-transition-duration: 0.1s;
  transition-duration: 0.1s;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  padding: 1rem;
  -webkit-box-pack: space-evenly;
  -webkit-justify-content: space-evenly;
  -ms-flex-pack: space-evenly;
  justify-content: space-evenly;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

<div>
  <div
    class="c0"
  >
    <button
      class="c1"
      type="button"
    >
      &lt; Previous
    </button>
    <div
      data-testid="pages-info"
    >
      Page 
      1
       of 
      100
    </div>
    <button
      class="c1"
      type="button"
    >
      Next &gt;
    </button>
  </div>
</div>
`;
